### PR TITLE
Fix sign and heaviside returning the wrong value for negative zero

### PR DESCRIPTION
**Pull request by AJ0070**
`ttnn.sign(-0.0)` returns `-1`. `torch.sign`, the golden registered for this op in
`ttnn/ttnn/operations/unary.py:71`, returns `0` for both signed zeros.

```
input   bits          ttnn.sign   torch.sign
-0.0    0x80000000    -1          0
 0.0    0x00000000     0          0
```

Reproduces on Blackhole p150b and Wormhole n150, in both `float32` and `bfloat16`.

## Cause

The SFPU comparison is not IEEE-compliant for signed zero. Measured on device with a
kernel that reports each predicate separately:

| input | `v == 0.0f` | `v < 0.0f` | `abs(v) == 0.0f` |
|---|---|---|---|
| `-0.0` | false | **true** | true |
| `+0.0` | true | false | true |

So `-0.0` reaches the `v < 0.0f` arm and is classified as negative. Note this also means
reordering the existing branches is not sufficient: `_sfpu_is_fp16_zero_` is `v == 0.0f`,
which is false for `-0.0`, so a zero-first ordering still falls through to the sign test.
Masking the sign bit with `abs()` is what matches both zeros.

## Fix

```c++
v_if(sfpi::abs(v) == 0.0F) { res = 0.0f; }
v_elseif(v < 0.0F) { res = -1.0f; }
v_endif;
```

Applied to both architecture copies, which were byte-identical. Quasar has no `sign`
kernel, so there is nothing to propagate there.

Negative zero is not exotic: it arises from subtracting equal values and from masked or
padded regions, so the wrong answer reaches ordinary graphs.

## Tests

Added `test_sign_signed_zero_fp32` and `test_sign_signed_zero_bfloat16` to
`tests/ttnn/unit_tests/operations/eltwise/test_unary_fp32.py`, which had no `sign`
coverage. Both fail without the fix and pass with it.

```
without the fix:  2 failed, 23 deselected
with the fix:     2 passed, 23 deselected
full file:        25 passed   (Blackhole p150b and Wormhole n150)
```

Non-zero inputs, both infinities and NaN are unchanged by this patch.

### CODE REVIEW

**Comment by AJ0070** [comment]
please re-run CI

File: tests/ttnn/unit_tests/operations/eltwise/test_unary_fp32.py [hunk 1]
```diff
@@ -294,3 +294,28 @@ def test_atanh(device, h, w):
...
+    assert torch.equal(tt_out, y_torch), f"got {tt_out.tolist()}, want {y_torch.tolist()}"
+
+
+@pytest.mark.parametrize("ttnn_function", [ttnn.sign])
+def test_sign_signed_zero_bfloat16(device, ttnn_function):
+    # Same defect on the bfloat16 path; the branch is shared.
+    x_torch = torch.tensor([[-0.0, 0.0, -1.0, 1.0, -5.5, 5.5]], dtype=torch.float32)
+    y_torch = torch.sign(x_torch)
+
+    x_tt = ttnn.from_torch(x_torch, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
```
[the remark this answers is not in the document]
Comment (AJ0070) [reply tests/ttnn/unit_tests/operations/eltwise/test_unary_fp32.py]: It does. Reverted the kernels and re-ran:

```
FAILED test_sign_signed_zero[dtype=bfloat16] - got [[-1.0, 0.0, ...]] want [[0.0, 0.0, ...]]
```

`negative_zero_delivered()` models the tt-llk matrix, where the format pair and `dest_acc` are set explicitly. ttnn resolves `unpack_to_dest_mode` differently for a bf16 tensor, so `-0.0` does reach the SFPU here.

File: tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_sign.h [hunk 2]
```diff
@@ -26,8 +26,10 @@ inline void calculate_sign(const uint /*exponent_size_8*/) {
     for (int d = 0; d < ITERATIONS; d++) {
         sfpi::vFloat v = sfpi::dst_reg[0];
         sfpi::vFloat res = 1.0f;
-        v_if(v < 0.0F) { res = -1.0f; }
-        v_elseif(_sfpu_is_fp16_zero_(v)) { res = 0.0f; }
+        // (v == 0.0f) is false for -0.0 on the SFPU while (v < 0.0f) is true, so
+        // _sfpu_is_fp16_zero_ misses negative zero. abs() matches both zeros.
+        v_if(sfpi::abs(v) == 0.0F) { res = 0.0f; }
```
[the remark this answers is not in the document]
Comment (AJ0070) [reply tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_sign.h:33]: Good catch, and heaviside is broken the same way: `heaviside(-0.0, 0.5)` returns 0 on baseline against torch 0.5, both dtypes. Fixed both here, since `_assert_signed_zero_partition_valid()` requires their sets to stay identical. The trailing `v_else` still catches NaN, so that is unchanged.

Left `_EDGE_KNOWN_DIVERGENCES` alone: those 4 combos now XPASS non-strictly. I cannot run the tt-llk suite here to confirm the partition still holds with both entries removed, so I would rather not rewrite that block blind. Happy to do it here if you want it in this PR.

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_sign.h [hunk 3]
```diff
@@ -26,8 +26,10 @@ inline void calculate_sign(const uint /*exponent_size_8*/) {
     for (int d = 0; d < ITERATIONS; d++) {
         sfpi::vFloat v = sfpi::dst_reg[0];
         sfpi::vFloat res = 1.0f;
-        v_if(v < 0.0F) { res = -1.0f; }
-        v_elseif(_sfpu_is_fp16_zero_(v)) { res = 0.0f; }
+        // (v == 0.0f) is false for -0.0 on the SFPU while (v < 0.0f) is true, so
```
[the remark this answers is not in the document]
Comment (AJ0070) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_sign.h]: Reworded to "unspecified", matching `ckernel_sfpu_binary.h:89`. Applied to both copies.

File: tests/ttnn/unit_tests/operations/eltwise/test_unary_fp32.py [hunk 4]
```diff
@@ -294,3 +294,28 @@ def test_atanh(device, h, w):
...
+def test_sign_signed_zero_fp32(device, ttnn_function):
+    # sign(-0.0) is 0, not -1. On the SFPU (v == 0.0f) is false for -0.0 while
+    # (v < 0.0f) is true, so the zero test has to mask the sign bit.
+    x_torch = torch.tensor([[-0.0, 0.0, -1.0, 1.0, -5.5, 5.5, -1e-30, 1e-30]], dtype=torch.float32)
+    y_torch = torch.sign(x_torch)
+
+    x_tt = ttnn.from_torch(x_torch, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device)
+    tt_out = ttnn.to_torch(ttnn_function(x_tt))
+
+    assert torch.equal(tt_out, y_torch), f"got {tt_out.tolist()}, want {y_torch.tolist()}"
```
[the remark this answers is not in the document]
Comment (AJ0070) [reply tests/ttnn/unit_tests/operations/eltwise/test_unary_fp32.py]: Done, and collapsed the two into one dtype-parametrised test per op, which drops the manual `.float()`. It was not in the import list, so added.

**Comment by AJ0070** [comment]
I have fixed the failing CI job, please re-run CI.

**Comment by ldjurovicTT** [comment]
@AJ0070 Great work!

Just one thing. In our tt-llk/tests/python_tests/test_eltwise_unary_sfpu.py we have a special xfail for this case. Please remove it from from there since you have fixed the error

**Comment by AJ0070** [comment]
Done, removed both xfails. I also fixed the signed-zero expectations in test_vif_equiv_sweep.py which were failing CI.

**Comment by AJ0070** [comment]
@ldjurovicTT #56138 makes the same sign and heaviside fix. This one also covers Blackhole, so it might be easier to land this first.

**Comment by ldjurovicTT** [comment]
@AJ0070 That is a draft of mine about all the remaining skipped special cases. You cover both arches, I think this one is also better to get merged first ( if mine is getting merged is still to be decided )

**Comment by AJ0070** [comment]
Thanks, that works for me. This one is ready on both arches whenever a reviewer can take it.

**Comment by AJ0070** [comment]
None of these look related. Clang tidy timed out at the 120 min cap, sdxl is the known GroupNorm perf flap in #56192, and the galaxy runner threw a PCIe read error before any test ran. Merged main in to retrigger.

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_heaviside.h [hunk 5]
```diff
@@ -25,7 +25,11 @@ inline void calculate_heaviside(uint value) {
...
 
-        v_if(v < 0.0f) { v = 0.0f; }
+        // SFPSETCC is unspecified for -0.0 (VectorUnit.md), so (v < 0.0f) can capture it
+        // and report 0 where -0.0 == 0 makes the answer s. SFPABS clears the sign bit.
+        // The trailing v_else still catches NaN, whose behaviour is unchanged.
+        v_if(sfpi::abs(v) == 0.0f) { v = s; }
+        v_elseif(v < 0.0f) { v = 0.0f; }
         v_elseif(v > 0.0f) { v = 1.0f; }
         v_else { v = s; }
         v_endif;
```
Comment (ldjurovicTT) [inline tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_heaviside.h]: can we use the copysign trick here also to make code shorter?

Comment (AJ0070) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_heaviside.h]: Done. NaN is unaffected, the old chain dispatched it on the sign bit too, so +nan still gives 1.0 and -nan 0.0 on both arches. Checked on p150b and n150.

**Comment by AJ0070** [comment]
The only failing check is the Blackhole simulator hitting an unimplemented tensix_stallwait in a DFB test this PR does not touch. Could someone re-run it?

**Comment by AJ0070** [comment]
@ldjurovicTT  CI is green!

**Review by ldjurovicTT — APPROVED** [review]

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_sign.h [hunk 6]
```diff
@@ -26,8 +26,9 @@ inline void calculate_sign(const uint /*exponent_size_8*/) {
     for (int d = 0; d < ITERATIONS; d++) {
         sfpi::vFloat v = sfpi::dst_reg[0];
         sfpi::vFloat res = 1.0f;
-        v_if(v < 0.0F) { res = -1.0f; }
-        v_elseif(_sfpu_is_fp16_zero_(v)) { res = 0.0f; }
+        // SFPSETCC is unspecified for -0.0 (VectorUnit.md), so a bare compare can miss it.
+        v_if(sfpi::abs(v) == 0.0F) { res = 0.0f; }
```
[the remark this answers is not in the document]
Comment (AJ0070) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_sign.h:32]: Removed from both sign copies in c41ffe39e3e.

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_heaviside.h [hunk 7]
```diff
@@ -25,12 +25,12 @@ inline void calculate_heaviside(uint value) {
     for (int d = 0; d < ITERATIONS; d++) {
         vFloat v = dst_reg[0];
 
-        v_if(v < 0.0f) { v = 0.0f; }
-        v_elseif(v > 0.0f) { v = 1.0f; }
-        v_else { v = s; }
+        vFloat r = sfpi::copysgn(vFloat(0.5f), v) + 0.5f;
+        // SFPSETCC is unspecified for -0.0 (VectorUnit.md), so a bare v == 0.0f can miss it.
+        v_if(sfpi::abs(v) == 0.0f) { r = s; }
```
[the remark this answers is not in the document]
Comment (AJ0070) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_heaviside.h:37]: Ported in c41ffe39e3e, citing the p150b before/after NaN check instead of the sweep.

File: tt_metal/tt-llk/tests/python_tests/test_vif_equiv_sweep.py [hunk 8]
```diff
@@ -181,19 +181,21 @@
...
+# pin that: sign(-0.0) -> -1 where IEEE says -0.0, and heaviside(-0.0) -> 0
+# where -0.0 == 0 gives the scalar. Both fell out of SFPSETCC, which
 # tt-isa-documentation specifies only "provided that VC is neither negative zero
-# nor any kind of NaN" -- so the sign-bit read on -0.0 is outside the
-# primitive's contract rather than a hardware fault. test_eltwise_unary_sfpu
-# carries the same two divergences as documented xfails; this test pins them
-# bit-exactly so a rewrite of these kernels cannot move them unnoticed.
+# nor any kind of NaN", so the sign-bit read on -0.0 was outside the primitive's
+# contract rather than a hardware fault. Both kernels now take the zero arm on
+# -0.0 by testing sfpi::abs(v), which is inside that contract, so those two rows
```
[the remark this answers is not in the document]
Comment (AJ0070) [reply tt_metal/tt-llk/tests/python_tests/test_vif_equiv_sweep.py]: Scoped both notes to -0.0 in c41ffe39e3e.

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_sign.h [hunk 9]
```diff
@@ -26,8 +26,9 @@ inline void calculate_sign(const uint /*exponent_size_8*/) {
...
+        v_elseif(v < 0.0F) { res = -1.0f; }
```
[the remark this answers is not in the document]
Comment (AJ0070) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_sign.h]: Applied in 7a2a3c334b2, so BH sign now matches the WH body. I don't have a p150b available right now, so this one isn't run on BH. It relies on LT0 reading the sign bit there, which the old BH heaviside chain showed for +/-NaN on p150b, and rsqrt_compat.h notes the same for -0.0 on BH silicon.

**Comment by fvranicTT** [comment]
@AJ0070 could you rebase your branch? Thanks.

**Comment by AJ0070** [comment]
@fvranicTT Rebased!

**Comment by AJ0070** [comment]
Please re-run the CI

**Comment by AJ0070** [comment]
@fvranicTT CI is green!

**Review by mateusznowakTT — APPROVED** [review]
